stdlib/public/TensorFlow/CompilerRuntime.swift

```diff
@@ -1507,6 +1512,58 @@ func _TFCOpSetAttrStringArray(_ op: CTFEOp,
   }
 }
 
+/// A TensorFlow device kind.
+public enum DeviceKind : UInt {
+  /// Default device.
+  case `default` = 0
+  /// Central processing units.
+  case cpu = 1
+  /// Graphics processing units.
+  case gpu = 2
+  // TODO: TPU?
+}
+
+internal extension _ExecutionContext {
+  @usableFromInline
+  func pushDeviceScope(_ kind: DeviceKind) {
+    sync {
+      deviceScopes.append(kind)
+    }
+  }
+
+  @usableFromInline
+  func popDeviceScope() {
+    sync {
+      internalConsistencyCheck(deviceScopes.popLast() != nil)
+    }
+  }
```

Comment:
Originally, I made `sync` private because I expected `_ExecutionContext` to only expose APIs that are thread-safe so that clients would not need to use sync. So, instead of making `sync(execute:)` internal, I think it's better to add the following internal computed property:

```swift
var currentDeviceKind: DeviceKind {
  return sync {
    deviceScopes.last ?? .default
  }
}
```

Then the runtime entry point will call this computed property to the current device instead of manually sync'ing threads. `sync(execute:)` will stay private.